### PR TITLE
Several cloud provider upgrades & enhancements

### PR DESCRIPTION
**Pull request by dstockton**
What
----

Improvement - Upgrade to golang 1.12
Improvement - Upgrade github.com/Azure/azure-sdk-for-go
Improvement - Upgrade github.com/aws/aws-sdk-go
Feature - AKS - Support for auto-scaling
Feature - EKS - Placement group support
Feature - EKS - Option to manage security groups outside of Rancher (e.g. when wrapped with Terraform)
Feature - EKS - Separate ASGs per subnet (for improved auto-scaling)
Feature - EKS - Option to use provided SSH key
Feature - GKE - Multi-zone (aka regional) cluster support
Feature - GKE - Control of DefaultMaxPodsConstraint (to better utilize IP address space)
Feature - AKS/EKS/GKE - Tag cloud resources with cluster annotations
BugFix - EKS - STS token handling (e.g. update creds after expired tokens)
BugFix - EKS - Fix broken k8s version update handling
BugFix - GKE - Fix broken k8s version update handling
BugFix - Better map argument type handling (so that `norman` validation doesn't force `nil` inputs over API interface)

### CODE REVIEW

**Comment by dstockton** [comment]
https://github.com/rancher/rancher/pull/24095 is the related Rancher PR

**Comment by leodotcloud** [comment]
One quick suggestion, could you separate your actual changes from go module changes? We generally change `go.mod`, run `go mod vendor` and commit those changes, followed by a commit with the actual changes.

**Comment by dstockton** [comment]
@leodotcloud I can if that would help but the intermediate stage would be a broken codebase wouldn’t it?

**Comment by dramich** [comment]
@dstockton No, it would all be merged at once. Looking at this breaking it into multiple commits with each cloud provider would also be helpful when reviewing

**Comment by dstockton** [comment]
@dramich. Yes, it’s a bit much to review in one go, I agree. I’ll see how I can best split it up.

**Comment by dstockton** [comment]
@dramich @leodotcloud - I've now re-worked the commits so it's much easier to review.
It'd be great to get some more feedback so that I can help make any changes necessary for this PR to be accepted.

Many thanks,
David

File: Dockerfile.dapper [hunk 1]
```diff
@@ -1,6 +1,6 @@
-FROM golang:1.10
+FROM golang:1.12
 RUN apt-get update && \
-    apt-get install -y xz-utils zip rsync
+    apt-get install -y xz-utils zip rsync musl-dev
```
Comment (leodotcloud) [inline Dockerfile.dapper]: We had faced some issues with `musl` in the past. Curious, what are you using this for?

File: drivers/aks/aks_driver.go [hunk 2]
```diff
@@ -114,6 +114,22 @@ type state struct {
 	// Tags tag the cluster. [optional]
 	Tags map[string]string `json:"tags,omitempty"`
 
+	/**
+	Auto-scaling related properties
```
Comment (leodotcloud) [inline drivers/aks/aks_driver.go]: Does it make sense to have a struct here?

File: Dockerfile.dapper (hunk 1 above)
Comment (dstockton) [reply Dockerfile.dapper]: I found the build was not successful without it on golang 1.12. I’ve been running it in production for a month or so without issue but equally have no specific requirements for it with the code changes I’ve made.

File: drivers/aks/aks_driver.go (hunk 2 above)
Comment (dstockton) [reply drivers/aks/aks_driver.go]: Sorry, could you elaborate on what you’re suggesting?

File: Dockerfile.dapper (hunk 1 above)
Comment (dstockton) [reply Dockerfile.dapper]: @leodotcloud I've done some more testing and it's not necessary - I've removed it

File: drivers/aks/aks_driver.go (hunk 2 above)
Comment (dstockton) [reply drivers/aks/aks_driver.go]: If you're suggesting putting EnableAutoScaling, MinCount,MaxCount, AgentPoolType, AvailabilityZones & LoadBalancerSku into a sub-struct; I could but was keeping close to the existing AKS & implementations - though I do note that GKE does use some `*raw` structs.

Note that AgentPoolType, AvailabilityZones & LoadBalancerSku are not necessarily limited to only when auto-scaling is enabled.

I don't feel it needs splitting into a sub-structure - but I'm eager to do what is necessary to get this functionality included in rancher - as the experts/owners, I'm happy for you guys to make a call on this.

Your time reviewing is much appreciated & I know it's often a thankless & time-consuming task - Thanks!

**Comment by dstockton** [comment]
@leodotcloud @dramich - if you have any further comments/concerns it'd be great to hear them - am eager to help get this merged.

**Comment by cjellick** [comment]
@dstockton thanks a lot for investing the time into this. Its really awesome. This PR is huge. Is there possibility that you can break it down into multiple smaller PRs that can be merged on their own?
I'd suggest breaking it down into the smallest functional units possible. The reality of the situation is that all the changes you are proposing are going to consume significant QA resources, so we may have to pick and choose which ones we can fit into the release cycle

**Comment by cjellick** [comment]
Also, it would help us expedite the process and make your changes more likely to get merged if you opened an issue for each functional change that your are making and link to it in the PR. Our QA process is powered by assigning GitHub issues over to QA for testing. This is how  they understand what needs to be tested.

**Comment by dstockton** [comment]
Per discussions, I've raised individual issues for items related to this PR...

Feature - AKS - Auto-Scaling
https://github.com/rancher/rancher/issues/24278

Bug - Additional cluster store validation (to prevent panic)
https://github.com/rancher/rancher/issues/24279

Bug - kubecfg only works when rancher is available on https://localhost
https://github.com/rancher/rancher/issues/24280

Feature - AKS/EKS/GKE - Tag cloud resources with cluster annotations
https://github.com/rancher/rancher/issues/24281

Bug - StringMapType support (so norman validation doesn’t clobber defaults to nil)
https://github.com/rancher/rancher/issues/24282

Feature - EKS - Placement group support
https://github.com/rancher/rancher/issues/24284

Feature - EKS - Option to manage security groups outside of Rancher (e.g. when wrapped with Terraform)
https://github.com/rancher/rancher/issues/24285

Feature - EKS - Separate ASGs per subnet (for AZ aware auto-scaling)
https://github.com/rancher/rancher/issues/24286

Feature - EKS - Option to use provided SSH key
https://github.com/rancher/rancher/issues/24287

Feature - GKE - Multi-zone (aka regional) cluster support
https://github.com/rancher/rancher/issues/24288

Feature - GKE - Control of DefaultMaxPodsConstraint (to better utilize IP address space)
https://github.com/rancher/rancher/issues/24289

Bug - EKS - STS token handling (e.g. update creds after expired tokens)
https://github.com/rancher/rancher/issues/24290

Bug - EKS - Fix broken k8s version update handling
https://github.com/rancher/rancher/issues/24291

Bug - GKE - Fix broken k8s version update handling
https://github.com/rancher/rancher/issues/24292

Bug - EKS - Typo in clusterDriverEKS (terraform provider)
https://github.com/rancher/rancher/issues/24293

Improvement - EKS - SessionToken marked as private credential
https://github.com/rancher/rancher/issues/24276

BugFix - EKS - eks.%.amazonaws.com/eks.%.amazonaws.com.cn added to domain whitelist
https://github.com/rancher/rancher/issues/24277

File: drivers/aks/aks_driver.go [hunk 3]
```diff
@@ -311,11 +331,58 @@ func (d *Driver) GetDriverCreateOptions(ctx context.Context) (*types.DriverFlags
...
 		Type:  types.StringSliceType,
+		Usage: "Availability zones to setup agent pools for.",
+		Default: &types.Default{
+			DefaultStringSlice: &types.StringSlice{Value: []string{"1"}}, //avoid nil value for init
+		},
+	}
+	driverFlag.Options["load-balancer-sku"] = &types.Flag{
+		Type:  types.StringType,
+		Usage: "Load balancer SKU (Basic or Standard).",
+		Value: "Basic",
```
Comment (loganhz) [inline drivers/aks/aks_driver.go]: This should be `basic` I think. It is lower-case in model of azure sdk.

File: drivers/aks/aks_driver.go [hunk 4]
```diff
@@ -752,10 +896,16 @@ func (d *Driver) createOrUpdate(ctx context.Context, options *types.DriverOption
 			driverState.Subnet,
 		))
 
-		networkProfile = &containerservice.NetworkProfile{
+		loadBalancerSku := containerservice.Basic
+		if driverState.LoadBalancerSku != "" {
```
Comment (loganhz) [inline drivers/aks/aks_driver.go]: If we add this code block here, it means only users select `Advanced` Network mode then they can choose the type of loadBalancerSku. However, in Azure console, users can use Standard LB with default Network mode.

So we might need to support Standard LB with default Network in Rancher as well

File: drivers/aks/aks_driver.go [hunk 5]
```diff
@@ -795,27 +945,58 @@ func (d *Driver) createOrUpdate(ctx context.Context, options *types.DriverOption
 			osDiskSizeGBPointer = to.Int32Ptr(int32(driverState.AgentOsdiskSizeGB))
 		}
 
-		agentStorageProfile := containerservice.ManagedDisks
-		if driverState.AgentStorageProfile != "" {
```
Comment (loganhz) [inline drivers/aks/aks_driver.go]: `agentStorageProfile` is not supported any more in the creating AKS API. I think this is the reason why you removed it here. Maybe we should clean it up such as https://github.com/rancher/kontainer-engine/pull/200/files#diff-fdee2f189d794fc3e135129d93b4d373L238

File: drivers/aks/aks_driver.go (hunk 3 above)
Comment (dstockton) [reply drivers/aks/aks_driver.go]: Resolved :+1:

File: drivers/aks/aks_driver.go (hunk 4 above)
Comment (dstockton) [reply drivers/aks/aks_driver.go]: Resolved :+1:

File: drivers/aks/aks_driver.go (hunk 5 above)
Comment (dstockton) [reply drivers/aks/aks_driver.go]: Resolved :+1:

**Pull request closed without merging**